src/python/pants/backend/tools/semgrep/rules.py

```diff
@@ -0,0 +1,232 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import itertools
+import logging
+from collections import defaultdict
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Iterable
+
+from pants.backend.python.util_rules import pex
+from pants.backend.python.util_rules.pex import PexRequest, VenvPex, VenvPexProcess
+from pants.core.goals.lint import LintResult, LintTargetsRequest
+from pants.core.util_rules.partitions import Partition, Partitions
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import (
+    CreateDigest,
+    Digest,
+    FileContent,
+    MergeDigests,
+    PathGlobs,
+    Paths,
+    Snapshot,
+)
+from pants.engine.process import FallibleProcessResult, ProcessCacheScope
+from pants.engine.rules import Get, MultiGet, Rule, collect_rules, rule
+from pants.engine.unions import UnionRule
+from pants.option.global_options import GlobalOptions
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+from .subsystem import SemgrepFieldSet, SemgrepSubsystem
+
+logger = logging.getLogger(__name__)
+
+
+class SemgrepLintRequest(LintTargetsRequest):
+    field_set_type = SemgrepFieldSet
+    tool_subsystem = SemgrepSubsystem
+
+
+@dataclass(frozen=True)
+class PartitionMetadata:
+    config_files: frozenset[Path]
+    ignore_files: Snapshot
+
+    @property
+    def description(self) -> str:
+        return ", ".join(sorted(str(path) for path in self.config_files))
+
+
+_IGNORE_FILE_NAME = ".semgrepignore"
+
+_RULES_DIR_NAME = ".semgrep"
+_RULES_FILES_GLOBS = (
+    ".semgrep.yml",
+    ".semgrep.yaml",
+    f"{_RULES_DIR_NAME}/*.yml",
+    f"{_RULES_DIR_NAME}/*.yaml",
+)
+
+
+@dataclass
+class SemgrepIgnoreFiles:
+    snapshot: Snapshot
+
+
+@dataclass
+class AllSemgrepConfigs:
+    configs_by_dir: dict[Path, set[Path]]
+
+    def ancestor_configs(self, address: Address) -> Iterable[Path]:
+        # TODO: introspect the semgrep rules and determine which (if any) apply to the files, e.g. a
+        # Python file shouldn't depend on a .semgrep.yml that doesn't have any 'python' or 'generic'
+        # rules, and similarly if there's path inclusions/exclusions.
+        # TODO: this would be better as actual dependency inference (e.g. allows inspection, manual
+        # addition/exclusion), but that can only infer 'full' dependencies and it is wrong (e.g. JVM
+        # things break) for real code files to depend on this sort of non-code linter config; requires
+        # dependency scopes or similar (https://github.com/pantsbuild/pants/issues/12794)
+        spec = Path(address.spec_path)
```

Comment:
Should this be `PurePath`? Generally we prefer the `Pure` variants since rules shouldn't be doing file I/O

Reply:
I don't think this got resolved?

It's the only one I feel strongly about.

Admittedly we should issue a lint.

Reply:
Oh, dear, sorry, I fixed but pushed to the wrong branch! Thanks for flagging.

https://github.com/pantsbuild/pants/pull/19685

Reply:
> Admittedly we should issue a lint.

https://github.com/pantsbuild/pants/pull/19686